### PR TITLE
update!: support all libscap events in the plugin system (syscalls included)

### LINKED ISSUE #992: [New Feature] Supporting syscall event source in plugin API
**Motivation**

The plugin system has been launched with the intention of extending the libraries by accessing a controlled subset of their capabilities. Currently, the plugin API only allows implementing and consuming extra event sources, such as the one of K8S audit and AWS cloudtrail events. However, one of the biggest missing points which have been requested by the community since the launch of the feature, is the ability to write plugins that deal with system calls. 

**Feature**

Extending the plugin API to support:
- Plugins with event sourcing capability being able to produce any of the events supported by libscap, and not just `PPME_PLUGIN_EVENT_E`
- Plugins with field extraction capability being able to implement field extractors from any of the events produced by libscap
- Potentially introducing a new capability for the generation of "meta-events", which are a currently non-standardized type of events generated internally in the libraries

**Additional context**

- https://github.com/falcosecurity/plugins/issues/85
- https://github.com/falcosecurity/falco/issues/1704
- https://kubernetes.slack.com/archives/CMWH3EH32/p1644507264083289
- https://kubernetes.slack.com/archives/CMWH3EH32/p1643838967857299

### PR DESCRIPTION
**Pull request by jasondellaluce**
**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

/area libscap-engine-source-plugin

/area libsinsp

/area tests

**Does this PR require a change in the driver versions?**

**What this PR does / why we need it**:

This PRs attempts solving the problem stated in #992. With this, the plugin framework complies with the event definitions of libscap for both reading and writing, thus effectively enabling:
- plugins producing any libscap event, just like our drivers always did (including syscalls)
- plugins extracting new fields from any libscap event (including syscalls)
- doing both of the above
- support all the already existing plugin features that enable defining new event sources and extracting data from them.

Unfortunately, this requires a major version change in the plugin API in order to avoid having both the old and new event definitions in the API.

This also tries to add unit tests for the event sourcing and field extraction capability, in many default and corner cases.

**Which issue(s) this PR fixes**:

Fixes #992

**Special notes for your reviewer**:

Please don't be scared of the LOC, 850+ of them are related to testing. For the core change proposition, disregard anything under `userspace/libsinsp/test` during the review.

**Does this PR introduce a user-facing change?**:

```release-note
update!: support all libscap events in the plugin system (syscalls included)
```

### CODE REVIEW

File: userspace/libscap/engine/source_plugin/plugin_info.h [hunk 1]
```diff
@@ -32,6 +32,6 @@ typedef struct
 
 	ss_instance_t* (*open)(ss_plugin_t* s, const char* params, ss_plugin_rc* rc);
 	void (*close)(ss_plugin_t* s, ss_instance_t* h);
-	ss_plugin_rc (*next_batch)(ss_plugin_t* s, ss_instance_t* h, uint32_t *nevts, ss_plugin_event **evts);
+	ss_plugin_rc (*next_batch)(ss_plugin_t* s, ss_instance_t* h, uint32_t *nevts, ss_plugin_event ***evts);
```
Comment (FedeDP) [inline userspace/libscap/engine/source_plugin/plugin_info.h:35]: Oh God :laughing:

File: userspace/libscap/engine/source_plugin/source_plugin.c [hunk 2]
```diff
@@ -32,6 +33,31 @@ static const char * const source_plugin_counters_stats_names[] = {
 	[N_EVTS] = "n_evts",
 };
 
+// We need to check that ppm_evt_hdr and ss_plugin_event are the same struct
+// right at compile time. We do so by thecking for their size and the offset
+// of each of their subfields. This allows us to avoid divergences while at the
+// same time not sharing the same headers.
+#if defined __GNUC__ || __STDC_VERSION__ >= 201112L
+_Static_assert(
```
Comment (FedeDP) [inline userspace/libscap/engine/source_plugin/source_plugin.c:41]: Thanks for this!

File: userspace/libscap/engine/source_plugin/source_plugin.c [hunk 3]
```diff
@@ -32,6 +33,31 @@ static const char * const source_plugin_counters_stats_names[] = {
 	[N_EVTS] = "n_evts",
 };
 
+// We need to check that ppm_evt_hdr and ss_plugin_event are the same struct
+// right at compile time. We do so by thecking for their size and the offset
```
Comment (FedeDP) [inline userspace/libscap/engine/source_plugin/source_plugin.c:37]:
```suggestion
// right at compile time. We do so by checking for their size and the offset
```

File: userspace/libsinsp/test/plugins/syscall_source.cpp [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    *evts = &istate->evt;
+    istate->evt->type = PPME_SYSCALL_OPEN_X;
+    istate->evt->tid = 1;
+    istate->evt->ts = UINT64_MAX;
+    istate->evt->len = sizeof(ss_plugin_event);
+    istate->evt->nparams = 6;
+
+    uint8_t* parambuf = &istate->evt_buf[0] + sizeof(ss_plugin_event);
+
+    // lenghts
```
Comment (FedeDP) [inline userspace/libsinsp/test/plugins/syscall_source.cpp:119]: Is the SDK going to provide helpers for this? 
Or do we expect that users wanting to send syscall events from a plugin know what they're doing (aka the contract, aka the event table and scap internals)?

**Review by FedeDP — COMMENTED** [review]
LGTM!
This is an incredible job, thank you @jasondellaluce !

File: userspace/libsinsp/test/plugins/syscall_source.cpp (hunk 4 above)
Comment (jasondellaluce) [reply userspace/libsinsp/test/plugins/syscall_source.cpp:119]: I see the scap specific to become more open and available to developers in the future now that's part of the plugin system, but of course the SDKs will take care of doing the extra lift in that case. From the Go SDK for example, I expect the UX to not change at all.

File: userspace/libscap/engine/source_plugin/plugin_info.h (hunk 1 above)
Comment (jasondellaluce) [reply userspace/libscap/engine/source_plugin/plugin_info.h:35]: Yeah, I know 🙄

Comment (leogr) [reply userspace/libscap/engine/source_plugin/plugin_info.h:35]: Has the explanation for that been documented somewhere? :thinking:

Comment (jasondellaluce) [reply userspace/libscap/engine/source_plugin/plugin_info.h:35]: In `plugin_api.h`.

**Review by FedeDP — DISMISSED** [review]
/approve
LGTM, incredible work! Thank you Jason!

**Review by FedeDP — APPROVED** [review]
/approve

**Review by leogr — APPROVED** [review]
:star_struck:
